### PR TITLE
ebpf : introduce a constant flow polling rate from the kernel

### PR DESCRIPTION
**Pull request by nplanel**
Avoid to burst load the kernel every tick
Default flow polling rate : 16000 flow per second

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -105,6 +105,7 @@ func init() {
 	cfg.SetDefault("agent.flow.pcapsocket.min_port", 8100)
 	cfg.SetDefault("agent.flow.pcapsocket.max_port", 8132)
 	cfg.SetDefault("agent.flow.ebpf.kernel_scan_interval", 10)
```
Comment (safchain) [inline config/config.go]: we need to find a way to convert this one to the new one, config relocation ?

File: config/config.go [hunk 2]
```diff
@@ -105,6 +105,7 @@ func init() {
...
+	cfg.SetDefault("agent.flow.ebpf.polling_rate", 1000000)
```
Comment (safchain) [inline config/config.go]: maybe better `flow_per_minute`
You need to modify the default config file too, will be nice to add a comment to elaborate about the usage of this

File: flow/probes/ebpf.go [hunk 3]
```diff
@@ -74,7 +74,7 @@ func (p *EBPFProbe) run() {
 	defer p.flowTable.Stop()
 
 	ebpfUpdate := time.Duration(p.Ctx.Config.GetInt("agent.flow.ebpf.kernel_scan_interval")) * time.Second
-	flowTableSz := p.Ctx.Config.GetInt("flow.max_entries")
+	ebpfConstantPollingRate := time.Minute / p.Ctx.Config.GetInt("agent.flow.ebpf.polling_rate")
```
Comment (safchain) [inline flow/probes/ebpf.go]: nit: just `ebpfPollingRate`

File: flow/probes/ebpf.go [hunk 4]
```diff
@@ -74,7 +74,7 @@ func (p *EBPFProbe) run() {
 	defer p.flowTable.Stop()
 
 	ebpfUpdate := time.Duration(p.Ctx.Config.GetInt("agent.flow.ebpf.kernel_scan_interval")) * time.Second
```
Comment (safchain) [inline flow/probes/ebpf.go]: We need to remove this as well no ?

**Comment by nplanel** [comment]
@eranra

**Comment by nplanel** [comment]
@eranra
To avoid burst kernel pressure every 10 seconds, we prefer the solution of a long smooth polling kernel flow table. This have impact to remove agent.flow.ebpf.kernel_scan_interval option.

**Comment by eranra** [comment]
@nplanel (1) This is fine, actually good idea/enhancement so no problem in dropping. What I try to think on, is if there is any implication on performance ... it feels like in both cases low and high stress this make sense ... I hope I am not missing anything here.

(2) why 'polling_rate' is per minute ... I think that per second will be less confusing

**Comment by safchain** [comment]
@eranra @nplanel I do agree per second will be better as we already have rate in this unit.

**Comment by nplanel** [comment]
@eranra 1/ The impact of this patch with high low, garantee a limited smooth kernel and userspace CPU load. Another hand flow will not be retrieve _quickly_ to the flow table, but as ebpf probe is about monitoring it's not really a problem.
2/ I changed the unit to flows per second (default : 16000)

**Comment by eranra** [comment]
@nplanel I get it ... all make sense (at least to me :-) )

File: config/config.go (hunk 2 above)
Comment (safchain) [reply config/config.go]: need to be adapted...

File: flow/probes/ebpf.go [hunk 5]
```diff
@@ -73,24 +73,29 @@ func (p *EBPFProbe) run() {
 	_, flowEBPFChan, _ := p.flowTable.Start()
 	defer p.flowTable.Stop()
 
-	ebpfUpdate := time.Duration(p.Ctx.Config.GetInt("agent.flow.ebpf.kernel_scan_interval")) * time.Second
-	flowTableSz := p.Ctx.Config.GetInt("flow.max_entries")
+	ebpfPollingRate := time.Second / p.Ctx.Config.GetInt("agent.flow.ebpf.polling_rate")
```
Comment (safchain) [inline flow/probes/ebpf.go]: unit test are failing because of this line

File: flow/probes/ebpf.go [hunk 6]
```diff
@@ -73,24 +73,31 @@ func (p *EBPFProbe) run() {
...
 	var nextAvailablePtr int
+	var now time.Time
 	for {
 		select {
-		case now := <-updateTicker.C:
+		case <-p.quit:
+			return
+		case now = <-updateNow.C:
+		default:
+			if now.IsZero() {
```
Comment (safchain) [inline flow/probes/ebpf.go]: what about now := time.Now() line 91 this will avoid this if each loop

Comment (nplanel) [reply flow/probes/ebpf.go]: yep much better solution

**Comment by nplanel** [comment]
run skydive-functional-hw-tests

**Comment by nplanel** [comment]
run skydive-functional-hw-tests

**Comment by safchain** [comment]
run skydive-functional-tests-backend-elasticsearch
run skydive-functional-hw-tests

File: tests/flow_test.go [hunk 7]
```diff
@@ -2311,9 +2311,6 @@ func testFlowsHashCnx(t *testing.T) *Test {
 			foundAB := false
 			foundBA := false
 			for _, f := range flows {
-				if f.Metric.ABPackets != 1 || f.Metric.BAPackets != 1 {
```
Comment (safchain) [inline tests/flow_test.go]: why removing this ?

**Comment by nplanel** [comment]
run skydive-functional-hw-tests

**Comment by nplanel** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by nplanel** [comment]
run skydive-functional-tests-backend-elasticsearch

**Comment by nplanel** [comment]
run skydive-functional-hw-tests

**Comment by nplanel** [comment]
run skydive-functional-hw-tests

**Comment by nplanel** [comment]
run skydive-functional-hw-tests

**Comment by nplanel** [comment]
run skydive-functional-hw-tests

**Comment by nplanel** [comment]
run skydive-functional-tests-backend-elasticsearch

**Comment by nplanel** [comment]
run skydive-functional-tests-backend-orientdb

**Pull request merged**